internal/config/pipeline.go

```diff
@@ -0,0 +1,283 @@
+func (c *Config) Pipelines() []string {
+	var result []string
+	for _, p := range c.hclConfig.Pipelines {
+		result = append(result, p.Name)
+	}
+
+	return result
+}
+
+// Pipeline returns the configured pipeline named n. If the pipeline doesn't
+// exist, this will return (nil, nil).
+func (c *Config) Pipeline(id string, ctx *hcl.EvalContext) (*Pipeline, error) {
+	ctx = appendContext(c.ctx, ctx)
+
+	// Find the pipeline by progressively decoding
+	var rawPipeline *hclPipeline
+	for _, p := range c.hclConfig.Pipelines {
+		if p.Name == id {
+			rawPipeline = p
+			break
+		}
+	}
+	if rawPipeline == nil {
+		return nil, nil
+	}
+
+	// Full decode
+	var pipeline Pipeline
+	if diag := gohcl.DecodeBody(rawPipeline.Body, finalizeContext(ctx), &pipeline); diag.HasErrors() {
+		return nil, diag
+	}
+	pipeline.Name = rawPipeline.Name
+	pipeline.ctx = ctx
+	pipeline.Config = c
+	if pipeline.Config != nil {
+		pipeline.Config.ctx = ctx
+	}
+
+	// decode all of the defined raw steps for a pipeline
+	var steps []*Step
+	for _, stepRaw := range pipeline.StepRaw {
+		// turn stepRaw into a staged Step
+		s := Step{
+			ctx:       ctx,
+			Name:      stepRaw.Name,
+			DependsOn: stepRaw.DependsOn,
+			ImageURL:  stepRaw.ImageURL,
+			Use:       stepRaw.Use,
+		}
+
+		steps = append(steps, &s)
+	}
+	pipeline.Steps = steps
+
+	return &pipeline, nil
+}
+
+// PipelineProtos will take the existing HCL eval context, eval the config
+// and translate the HCL result into a Pipeline Proto to be returned for
+// operations such as ConfigSync.
+func (c *Config) PipelineProtos() ([]*pb.Pipeline, error) {
+	if c == nil {
+		// This is likely an internal error if this happens.
+		panic("attempted to construct pipeline proto on a nil genericConfig")
+	}
+
+	// Load HCL config and convert to a Pipeline proto
+	var result []*pb.Pipeline
+	for _, pl := range c.hclConfig.Pipelines {
+		pipe := &pb.Pipeline{
+			Name: pl.Name,
+			Owner: &pb.Pipeline_Project{
+				Project: &pb.Ref_Project{
+					Project: c.hclConfig.Project,
+				},
+			},
+		}
+
+		// TODO(briancain): Use the default ODR image for image_url so that it isn't required
+		// We do this already maybe in the ODR task launcher.
+		steps := make(map[string]*pb.Pipeline_Step)
+		for i, step := range pl.StepRaw {
+			s := &pb.Pipeline_Step{
+				Name:      step.Name,
+				DependsOn: step.DependsOn,
+				Image:     step.ImageURL, //TODO(briancain): actually use this when executing steps
+			}
+
+			// If no dependency was explictily set, we rely on the previous step
+			if i != 0 && len(step.DependsOn) == 0 {
+				s.DependsOn = []string{pl.StepRaw[i-1].Name}
+			}
+
+			// We currently only support one kind of Step plugin. But in the future
+			// maybe this would be a switch on step.Type? Or maybe we get our
+			// own Step Eval func that returns the step proto instead and does the
+			// switches there.
+			// NOTE(briancain): This is what you'd change to support future Step plugins
+			switch step.Use.Type {
```

Comment:
⏳ I think in the future when running pipelines locally works (right now they work with remote ops only), we could also include the registry step here to run a push op. It's not necessary right now as it is since the build operation should also push in remote operations, so this is of course not blocking, but it came to mind when thinking about supporting each of the stanzas within `app`. 😄

Reply:
I do think there's value in allowing for locally run pipelines, especially while you're working on building them. Let's track that as a future enhancement for sure!

We can include a stand-alone registry step in follow-up work. I think it might be a bit tricky because we'll need to know what artifact we're actually pushing and or allow for referencing prior steps to push prior artifacts.